### PR TITLE
ERC20WithPermit token

### PR DESCRIPTION
**Pull request by pdyraga**
Generic implementation of ERC20 token with permit functionality, used
already in [keep-network/tbtc-v2](https://github.com/keep-network/tbtc-v2) and [keep-network/coverage-pools](https://github.com/keep-network/coverage-pools/).

User can authorize a transfer of their token with a signature conforming
EIP712 standard instead of an on-chain transaction from their address.
Anyone can submit this signature on the user's behalf by calling `permit`
function, as specified in EIP2612, paying gas fees, and possibly
performing other actions in the same transaction.

By the time of implementing this contract, to my knowledge, there is no
other audited EP2612 token implementation other than Uniswap V2 on which
this code is based on. OpenZeppelin has quite recently added `permit` to
their ERC20 but this code is still in a draft state, was released in the
most recent version of the library, and to my understanding was not
audited yet by another firm, and could have breaking change in the
future.

The implementation mostly mirrors UniswapV2ERC20. It adds some tweaks
from OpenZeppelin ERC20 implementation such as more meaningful revert
messages and Approval event emitted on calls to `transferFrom`. This
allows applications to reconstruct the allowance for all accounts just
by listening to the event. Additionally, this token implements
`approveAndCall` pattern.

Unit tests were based on OpenZeppelin ERC20 unit tests with our own
tests implementation for EIP2612 and EIP712. The token code in  
[keep-network/tbtc-v2](https://github.com/keep-network/tbtc-v2) this is is a ctrlc/ctrlv was audited by ToB.

### CODE REVIEW

File: contracts/token/ERC20WithPermit.sol [hunk 1]
```diff
@@ -2,6 +2,224 @@
...
+import "./IReceiveApproval.sol";
+
+/// @title  ERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
```
Comment (nkuba) [inline contracts/token/ERC20WithPermit.sol:17]: Don't we want to implement `IERC20Metadata` as OZ's `ERC20` does?

File: contracts/token/ERC20WithPermit.sol [hunk 2]
```diff
@@ -2,6 +2,224 @@
...
+
+    function buildDomainSeparator() private view returns (bytes32) {
+        return
+            keccak256(
+                abi.encode(
+                    keccak256(
+                        "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                    ),
+                    keccak256(bytes(name)),
+                    keccak256(bytes("1")),
```
Comment (nkuba) [inline contracts/token/ERC20WithPermit.sol:212]: Maybe we want to pull the version to a constant defined on top of this contract or to be configurable on contract deployment?

File: contracts/token/ERC20WithPermit.sol [hunk 3]
```diff
@@ -2,6 +2,224 @@
...
+        return true;
+    }
+
+    /* solhint-disable-next-line func-name-mixedcase */
+    function DOMAIN_SEPARATOR() public view override returns (bytes32) {
+        if (chainId() == cachedChainId) {
+            return cachedDomainSeparator;
+        } else {
+            return buildDomainSeparator();
+        }
```
Comment (nkuba) [inline contracts/token/ERC20WithPermit.sol:168]: I understand that this is a solution for security consideration mentioned in [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612):
 > If the DOMAIN_SEPARATOR contains the chainId and is defined at contract deployment instead of reconstructed for every signature, there is a risk of possible replay attacks between chains in the event of a fututre chain split.

Am I correct? If so, can we mention that in docs?

**Review by nkuba — COMMENTED** [review]
First comments round (without tests review).

File: contracts/token/IERC20WithPermit.sol [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    ///         from caller's allowance.
+    function burnFrom(address account, uint256 amount) external;
+
+    /// @notice Executes receiveApproval function on spender as specified in
+    ///         IReceiveApproval interface. Approves spender to withdraw from
+    ///         the caller multiple times, up to the value amount. If this
+    ///         function is called again, it overwrites the current allowance
+    ///         with value. Reverts if the approval reverted or if
+    ///         receiveApproval call on the spender reverted.
+    function approveAndCall(
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/IERC20WithPermit.sol:47]: Our `ERC20WithPermit` implements it. I was looking into extracting the implementation to another contract but then it is getting hairy because that contract needs to assume there is a base ERC20 contract that implements `approve`. And we start implementing a framework then... I think the gain is not worth the complexity.

File: contracts/token/ERC20WithPermit.sol [hunk 5]
```diff
@@ -2,6 +2,224 @@
 
 pragma solidity 0.8.4;
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol]: https://github.com/thesis/solidity-contracts/pull/3/commits/1228f10e34c9df1d44eccafd37182940b45e666a

File: contracts/token/ERC20WithPermit.sol (hunk 3 above)
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:175]: Yes! Done in https://github.com/thesis/solidity-contracts/pull/3/commits/8c4693de2747e4cad52ed24c9b4d75fd970e7232.

File: contracts/token/ERC20WithPermit.sol [hunk 6]
```diff
@@ -2,6 +2,224 @@
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:2]: I do not know, that's how it looks in all other contracts. 🤔

File: contracts/token/ERC20WithPermit.sol (hunk 1 above)
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:17]: See 6cba00526a60ba195d252e34c200d2be868b463b please.

File: contracts/token/IERC20WithPermit.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/IERC20WithPermit.sol]: Done in https://github.com/thesis/solidity-contracts/pull/3/commits/5596b66132ca76349036f57a29cb55c95651cc6a.

File: contracts/token/ERC20WithPermit.sol (hunk 2 above)
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:219]: This value is strictly related to EIP712 domain we are creating in this function. It is the current major version of the signing domain and the domain we are constructing is the domain in version 1. This version is not something we want to configure when deploying.

File: contracts/token/ERC20WithPermit.sol [hunk 7]
```diff
@@ -2,6 +2,224 @@
...
+
+/// @title  ERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
+contract ERC20WithPermit is IERC20WithPermit, Ownable {
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:18]: The goal is to have an audited version of the token with permit functionality and to do not care about version changes in referenced libraries. If we take the implementation from OpenZeppelin, then for every version upgrade we should inspect changes in the token. Separate from that, I am not sure about v4 but in v3 it was not possible, given that their `approve` function was not supporting `uint256.max` case specified in EIP2612 and we'd had to override `approve` which makes this construction complicated.

File: contracts/token/ERC20WithPermit.sol [hunk 8]
```diff
@@ -1,7 +1,232 @@
...
+        return true;
+    }
+
+    function transferFrom(
+        address sender,
+        address recipient,
+        uint256 amount
+    ) external override returns (bool) {
+        uint256 currentAllowance = allowance[sender][msg.sender];
+        if (currentAllowance != type(uint256).max) {
```
Comment (nkuba) [inline contracts/token/ERC20WithPermit.sol:61]: So if `currentAllowance` is equal `type(uint256).max` we won't get into this if statement. We won't decrease the allowance with the `_approve` function, but execute `_transfer` only. Is this some feature that allows unlimited tokens transfers? 
I couldn't find it in [OZ ERC20](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L149), is it a tweak of [UniswapV2ERC20 solution](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2ERC20.sol#L74)?
If this is an intentional feature we should mention it in the docs. 

By the way, we're missing docs for all the functions, so maybe it would be worth adding them?

File: contracts/token/IERC20WithPermit.sol [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
+interface IERC20WithPermit is IERC20, IERC20Metadata, IApproveAndCall {
+    /// @notice EIP2612 approval made with secp256k1 signature.
+    ///         Users can authorize a transfer of their tokens with a signature
+    ///         conforming EIP712 standard, rather than an on-chain transaction
+    ///         from their address. Anyone can submit this signature on the
+    ///         user's behalf by calling the permit function, paying gas fees,
+    ///         and possibly performing other actions in the same transaction.
+    /// @dev    The deadline argument can be set to uint(-1) to create permits
```
Comment (nkuba) [inline contracts/token/IERC20WithPermit.sol:25]: Is it deadline set to `-1` or the amount set to max uint256? Is it related to what I mentioned in my other comment around `currentAllowance != type(uint256).max`?

File: test/token/ERC20WithPermit.test.js [hunk 10]
```diff
@@ -0,0 +1,1069 @@
...
+            )
+        ).to.be.revertedWith("Approve to the zero address")
+      })
+    })
+
+    context("when given never expiring permit", () => {
+      // uint(-1)
+      const allowance = ethers.BigNumber.from(
+        "115792089237316195423570985008687907853269984665640564039457584007913129639935"
+      )
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:969]: Can we extract this value to a `MAX_UINT256` constant in `contract-test-helpers.js`?

File: test/token/ERC20WithPermit.test.js [hunk 11]
```diff
@@ -0,0 +1,1069 @@
...
+              signature.v,
+              signature.r,
+              signature.s
+            )
+        ).to.be.revertedWith("Approve to the zero address")
+      })
+    })
+
+    context("when given never expiring permit", () => {
+      // uint(-1)
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:966]: Is this still valid in solidity 0.8? We use max uint256 in the code, so having `uint(-1)` here is confusing.

File: test/token/ERC20WithPermit.test.js [hunk 12]
```diff
@@ -0,0 +1,1069 @@
...
+      const allowance = ethers.BigNumber.from(
+        "115792089237316195423570985008687907853269984665640564039457584007913129639935"
+      )
+
+      beforeEach(async () => {
+        const deadline = tomorrow
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:972]: According to the documentation `deadline` should be set to `unit(-1)`:
https://github.com/thesis/solidity-contracts/blob/5596b66132ca76349036f57a29cb55c95651cc6a/contracts/token/IERC20WithPermit.sol#L25-L26

But I believe we use `allowance` of max UINT256 for unlimited transfers:
https://github.com/thesis/solidity-contracts/blob/5596b66132ca76349036f57a29cb55c95651cc6a/contracts/token/ERC20WithPermit.sol#L61

File: contracts/token/ERC20WithPermit.sol (hunk 8 above)
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:61]: Correct. It is from [EIP-2612](https://eips.ethereum.org/EIPS/eip-2612):
> The deadline argument can be set to uint(-1) to create Permits that effectively never expire.

I don't think OpenZeppelin implements this part. I guess it's because EIP1612 implementation is still treated as a draft in their repo.

Comment (nkuba) [reply contracts/token/ERC20WithPermit.sol:61]: EIP-2612 is saying about unlimited time for the permit (deadline = uint(-1)), but here we do the unlimited allowed amount.

File: test/token/ERC20WithPermit.test.js (hunk 12 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js:972]: Yes, this is confusing, I'll rework it.

File: test/token/ERC20WithPermit.test.js (hunk 11 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js:966]: `uint(-1) = type(uint256).max`, see [Solidity 8 breaking changes](https://docs.soliditylang.org/en/v0.8.0/080-breaking-changes.html):

> In particular, replace usages of uint(-1) with type(uint).max.

I left it as uint(-1) because it's how it's referenced in the EIP.

File: test/token/ERC20WithPermit.test.js (hunk 10 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js:969]: We can but I I think it will be used only in this single test.

File: contracts/token/IERC20WithPermit.sol (hunk 9 above)
Comment (pdyraga) [reply contracts/token/IERC20WithPermit.sol:25]: Both, and yes, addressed it in another comment.

File: contracts/token/ERC20WithPermit.sol (hunk 8 above)
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:61]: Correct. EIP-2612 says:

> The deadline argument can be set to uint(-1) to create Permits that effectively never expire.

It says nothing about unit(-1) approvals but [Uniswap LP token](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2ERC20.sol#L74) mentioned as a reference implementation of EIP-2612 implements them. Same for [Maker DAO's DAI](https://github.com/makerdao/dss/blob/3f30552a586264b32ebdb5bac94ba67020282e53/src/dai.sol#L80). I think it makes sense given that permit and approval are siblings and if we can set unit(-1) deadline, why not to set unit(-1) approval. EIP-2612 is still in draft, so we'll have some small slight differences between tokens.

Regarding docs, we have them defined on the interface and EIP-20 is well-known so I don't think having them on the token would be a huge improvement.

If we want to add docs, I'd have [them the same way as OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/566a774222707e424896c0c390a84dc3c13bdcb2/contracts/token/ERC20/ERC20.sol#L124-L130) and just list subtle details, like unit(-1) approval amount. Does it work?

File: test/token/ERC20WithPermit.test.js (hunk 10 above)
Comment (nkuba) [reply test/token/ERC20WithPermit.test.js:969]: This can also serve as a helper that can be used in the other tests or repos, as this is the purpose of this repository. Plus having `MAX_UINT256` in the test instead of `115792089237316195423570985008687907853269984665640564039457584007913129639935` makes it clear what the value is.

File: contracts/token/ERC20WithPermit.sol (hunk 8 above)
Comment (nkuba) [reply contracts/token/ERC20WithPermit.sol:61]: > Regarding docs, we have them defined on the interface and EIP-20 is well-known so I don't think having them on the token would be a huge improvement.

I agree to the standard `ERC20` functions, but the `ERC20WithPermit` has some significant changes in `transferFrom` which definitely should be mentioned in the documentation.

Comment (nkuba) [reply contracts/token/ERC20WithPermit.sol:61]: > It says nothing about unit(-1) approvals but Uniswap LP token mentioned as a reference implementation of EIP-2612 implements them. Same for Maker DAO's DAI. I think it makes sense given that permit and approval are siblings and if we can set unit(-1) deadline, why not to set unit(-1) approval. EIP-2612 is still in draft, so we'll have some small slight differences between tokens.

So we will have two unlimited possibilities for `deadline` and `amount`.  We need to be clear in the documentation.
Also keeping mentioning `uint(-1)` is confusing as it's not reflected in the code as it's not supported in Solidity 0.8. We should prioritize mentioning max uint256 in the documentation with a brief reference that it replaces `uint(-1)`.

File: test/token/ERC20WithPermit.test.js (hunk 11 above)
Comment (nkuba) [reply test/token/ERC20WithPermit.test.js:966]: If we want to keep referencing code and EIP we should mention them both, something like:
```suggestion
      // uint(-1) = type(uint256).max
```

File: contracts/token/ERC20WithPermit.sol [hunk 13]
```diff
@@ -1,7 +1,232 @@
...
+            require(
+                currentAllowance >= amount,
+                "Transfer amount exceeds allowance"
+            );
+            _approve(sender, msg.sender, currentAllowance - amount);
+        }
```
Comment (nkuba) [inline contracts/token/ERC20WithPermit.sol:67]: Shouldn't we cover the same path in the `burnFrom` function? It also decreases allowance, so if `burnFrom` is called the allowance will drop below maxuint256, and will affect the expected functionality for transfers.

File: contracts/token/ERC20WithPermit.sol (hunk 8 above)
Comment (Shadowfiend) [reply contracts/token/ERC20WithPermit.sol:61]: I'm going to be a butt here: please let's add solid documentation, not half-assed docs like most contracts. Having to reference ERC20 to understand mechanics constantly is… Not great. Doesn't need to block any work, but we should look at each contract we formalize in our own code as an opportunity to raise the bar.

Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:90]: https://github.com/thesis/solidity-contracts/pull/3/commits/dec702f9f8fb532d4f1d83d8f6aa56e349b8f6b3

File: contracts/token/ERC20WithPermit.sol (hunk 13 above)
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol:96]: https://github.com/thesis/solidity-contracts/pull/3/commits/ee42c58081a345f759f8f89b7022c0f70109f14d

File: test/token/ERC20WithPermit.test.js (hunk 12 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js]: https://github.com/thesis/solidity-contracts/pull/3/commits/ee42c58081a345f759f8f89b7022c0f70109f14d

File: test/token/ERC20WithPermit.test.js (hunk 11 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js]: https://github.com/thesis/solidity-contracts/pull/3/commits/ee42c58081a345f759f8f89b7022c0f70109f14d

File: test/token/ERC20WithPermit.test.js (hunk 10 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js]: https://github.com/thesis/solidity-contracts/pull/3/commits/ee42c58081a345f759f8f89b7022c0f70109f14d

File: test/token/ERC20WithPermit.test.js [hunk 14]
```diff
@@ -0,0 +1,1102 @@
...
+  describe("balanceOf", () => {
+    context("when the requested account has no tokens", () => {
+      it("should return zero", async () => {
+        expect(await token.balanceOf(anotherAccount.address)).to.equal(0)
+      })
+    })
+
+    context("when the requested account has some tokens", () => {
+      it("should return the total amount of tokens", async () => {
+        expect(await token.balanceOf(initialHolder.address)).to.equal(
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:98]: Can we use another account that holds just a specific number of tokens and not the whole token supply?

File: test/token/ERC20WithPermit.test.js [hunk 15]
```diff
@@ -0,0 +1,1102 @@
...
+      const expected = ethers.utils.keccak256(
+        ethers.utils.toUtf8Bytes(
+          "Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)"
+        )
+      )
+      expect(await token.PERMIT_TYPEHASH()).to.equal(expected)
+    })
+  })
+
+  describe("domain_separator", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:63]:
```suggestion
  describe("DOMAIN_SEPARATOR", () => {
```

File: test/token/ERC20WithPermit.test.js [hunk 16]
```diff
@@ -0,0 +1,1102 @@
...
+          initialSupply
+        )
+      })
+    })
+  })
+
+  describe("transfer", () => {
+    context("when the recipient is not the zero address", () => {
+      context("when the sender does not have enough balance", () => {
+        const amount = initialSupply.add(1)
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:108]: Can we use another account that holds just a specific number of tokens and not the whole token supply?

File: test/token/ERC20WithPermit.test.js [hunk 17]
```diff
@@ -0,0 +1,1102 @@
...
+
+      context("when the sender transfers all balance", () => {
+        const amount = initialSupply
+
+        it("should transfer the requested amount", async () => {
+          await token.connect(initialHolder).transfer(recipient.address, amount)
+
+          expect(await token.balanceOf(initialHolder.address)).to.equal(0)
+
+          expect(await token.balanceOf(recipient.address)).to.equal(amount)
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:125]: Can we also validate that the `totalSupply` remained unchanged? This could be included in an `afterEach` block for all `transfer` tests.

```js
    afterEach(async () => {
      expect(await token.totalSupply()).to.equal(initialSupply)
    })
```

File: test/token/ERC20WithPermit.test.js [hunk 18]
```diff
@@ -0,0 +1,1102 @@
...
+        const amount = initialSupply.add(1)
+
+        it("should revert", async () => {
+          await expect(
+            token.connect(initialHolder).transfer(recipient.address, amount)
+          ).to.be.revertedWith("Transfer amount exceeds balance")
+        })
+      })
+
+      context("when the sender transfers all balance", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:117]: Can we add two test cases to cover:
- when the sender transfers an amount that is not the whole balance, so some tokens are left on their account,
- when sender sends two subsequent `transfer` events to the same recipient, to check that the balances were incremented and not reasigned to the latest tx value

File: contracts/token/ERC20WithPermit.sol [hunk 19]
```diff
@@ -1,7 +1,314 @@
...
+    /// @return True if the operation succeeded, reverts otherwise.
+    /// @dev Requirements:
+    ///       - `recipient` cannot be the zero address,
+    ///       - the caller must have a balance of at least `amount`.
+    function transfer(address recipient, uint256 amount)
+        external
+        override
+        returns (bool)
+    {
+        _transfer(msg.sender, recipient, amount);
```
Comment (nkuba) [inline contracts/token/ERC20WithPermit.sol:71]: Shouldn't we use `transferFrom` here? Aren't we missing allowance check right now?

File: test/token/ERC20WithPermit.test.js [hunk 20]
```diff
@@ -0,0 +1,1102 @@
...
+      it("should revert", async () => {
+        await expect(
+          token.connect(initialHolder).transfer(ZERO_ADDRESS, initialSupply)
+        ).to.be.revertedWith("Transfer to the zero address")
+      })
+    })
+  })
+
+  describe("transferFrom", () => {
+    context("when the token owner is not the zero address", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:174]: Can we add an oposite test case `when the token owner is the zero address`?

File: test/token/ERC20WithPermit.test.js [hunk 21]
```diff
@@ -0,0 +1,1102 @@
...
+        it("should revert", async () => {
+          await expect(
+            token
+              .connect(anotherAccount)
+              .transferFrom(initialHolder.address, ZERO_ADDRESS, allowance)
+          ).to.be.revertedWith("Transfer to the zero address")
+        })
+      })
+
+      context("when given the maximum allowance", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:340]: This should be moved under `context("when the recipient is not the zero address", () => {`.

File: test/token/ERC20WithPermit.test.js [hunk 22]
```diff
@@ -0,0 +1,1102 @@
...
+        ).to.be.revertedWith("Transfer to the zero address")
+      })
+    })
+  })
+
+  describe("transferFrom", () => {
+    context("when the token owner is not the zero address", () => {
+      context("when the recipient is not the zero address", () => {
+        context("when the spender has enough approved balance", () => {
+          const allowance = initialSupply
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:177]: Let's use amounts different than `initialSupply`.

File: test/token/ERC20WithPermit.test.js [hunk 23]
```diff
@@ -0,0 +1,1102 @@
...
+            const amount = initialSupply
+
+            it("should transfer the requested amount", async () => {
+              await token
+                .connect(anotherAccount)
+                .transferFrom(initialHolder.address, recipient.address, amount)
+
+              expect(await token.balanceOf(initialHolder.address)).to.equal(0)
+
+              expect(await token.balanceOf(recipient.address)).to.equal(amount)
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:194]: Let's check if `totalSupply` remained unchanged for all `transferFrom` tests.

File: test/token/ERC20WithPermit.test.js [hunk 24]
```diff
@@ -0,0 +1,1102 @@
...
+        context("when the spender has enough approved balance", () => {
+          const allowance = initialSupply
+          beforeEach(async function () {
+            await token
+              .connect(initialHolder)
+              .approve(anotherAccount.address, allowance)
+          })
+
+          context("when the token owner has enough balance", () => {
+            const amount = initialSupply
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:185]: Let's use amounts different than `initialSupply`, so we can compare that exactly the amount got transfered and tokens remained on sender account is a specific decreased amount, but not `0`.

File: contracts/token/IERC20WithPermit.sol [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+///         transaction.
+interface IERC20WithPermit is IERC20, IERC20Metadata, IApproveAndCall {
+    /// @notice EIP2612 approval made with secp256k1 signature.
+    ///         Users can authorize a transfer of their tokens with a signature
+    ///         conforming EIP712 standard, rather than an on-chain transaction
+    ///         from their address. Anyone can submit this signature on the
+    ///         user's behalf by calling the permit function, paying gas fees,
+    ///         and possibly performing other actions in the same transaction.
+    /// @dev    The deadline argument can be set to `type(uint256).max to create
+    ///         permits that effectively never expire.
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/IERC20WithPermit.sol:26]: uint(-1) deadline is mentioned in the EIP, I do not think it's an implementation detail.

File: test/helpers/contract-test-helpers.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
+function to1e18(n) {
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply test/helpers/contract-test-helpers.js:1]: Agreed. Right now we are happily copy-pasting them...

File: test/token/ERC20WithPermit.test.js [hunk 27]
```diff
@@ -0,0 +1,1102 @@
...
+                .to.emit(token, "Approval")
+                .withArgs(
+                  initialHolder.address,
+                  anotherAccount.address,
+                  allowance.sub(amount)
+                )
+            })
+          })
+
+          context("when the token owner does not have enough balance", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:235]: Can we add a test case where sender sends two subsequend `transferFrom` transctions to the same `recipient`?

File: test/token/ERC20WithPermit.test.js [hunk 28]
```diff
@@ -0,0 +1,1102 @@
...
+        const allowance = initialSupply
+
+        it("should emit an approval event", async () => {
+          const tx = await token
+            .connect(initialHolder)
+            .approve(anotherAccount.address, allowance)
+
+          await expect(tx)
+            .to.emit(token, "Approval")
+            .withArgs(initialHolder.address, anotherAccount.address, allowance)
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:378]: Can we also validate `allowance` value?

File: test/token/ERC20WithPermit.test.js (hunk 15 above)
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js]: https://github.com/thesis/solidity-contracts/pull/3/commits/d32f786c069b7aca0e48b493acfb9f7567acc611

File: test/token/ERC20WithPermit.test.js [hunk 29]
```diff
@@ -0,0 +1,1102 @@
...
+    expect(await token.decimals()).to.equal(18)
+  })
+
+  describe("totalSupply", () => {
+    it("should return the total amount of tokens", async () => {
+      expect(await token.totalSupply()).to.equal(initialSupply)
+    })
+  })
+
+  describe("permit_typehash", () => {
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply test/token/ERC20WithPermit.test.js]: https://github.com/thesis/solidity-contracts/pull/3/commits/d32f786c069b7aca0e48b493acfb9f7567acc611

File: contracts/token/ERC20WithPermit.sol [hunk 30]
```diff
@@ -1,7 +1,314 @@
...
+            require(
+                currentAllowance >= amount,
+                "Burn amount exceeds allowance"
+            );
+            _approve(account, msg.sender, currentAllowance - amount);
+        }
+        _burn(account, amount);
+    }
+
+    /// @notice Calls `receiveApproval` function on spender previusly approving
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol]: b61db77c98db5ed1907697e97bb00af27ab9a6ef

File: contracts/token/ERC20WithPermit.sol [hunk 31]
```diff
@@ -1,7 +1,314 @@
...
+    /// @notice Returns hash of EIP712 Domain struct with the token name as
+    ///         a signing domain and token contract as a verifying contract.
+    ///         Used to construct EIP2612 signature provided to `permit`
+    ///         function.
+    /* solhint-disable-next-line func-name-mixedcase */
+    function DOMAIN_SEPARATOR() public view override returns (bytes32) {
+        // As explained in EIP-2612, if the DOMAIN_SEPARATOR contains the
+        // chainId and is defined at contract deployment instead of
+        // reconstructed for every signature, there is a risk of possible replay
+        // attacks between chains in the event of a fututre chain split.
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply contracts/token/ERC20WithPermit.sol]: b61db77c98db5ed1907697e97bb00af27ab9a6ef

File: test/token/ERC20WithPermit.test.js [hunk 32]
```diff
@@ -0,0 +1,1102 @@
...
+      })
+    })
+  })
+
+  describe("approve", () => {
+    context("when the spender is not the zero address", () => {
+      context("when the sender has enough balance", () => {
+        const allowance = initialSupply
+
+        it("should emit an approval event", async () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:371]: I know it's validated in `"should approve the requested amount and replaces the previous one"` in another test but can we add `allowance` map check as well under `"when the sender has enough balance"`?

**Comment by nkuba** [comment]
Oh boy, I'm sorry for spamming. I was using VSC plugin for the review and I was 100% sure the comments are being added to the pending review so I will submit them all at once 🤦🏻

**Comment by pdyraga** [comment]
Let's open a separate issue for test improvements. All tests but permit-related were ported from OpenZeppelin and already reviewed in coverage-pools PR. If we want to do some modifications, let's not hold PRs that wait for this one.

File: test/token/ERC20WithPermit.test.js [hunk 33]
```diff
@@ -0,0 +1,1102 @@
...
+              deadline,
+              signature.v,
+              signature.r,
+              signature.s
+            )
+        ).to.be.revertedWith("Permission expired")
+      })
+    })
+
+    context("when permission has an invalid signature", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:740]: We should cover two revert cases:
- `Invalid signature 's' value`
- `Invalid signature 'v' value`

File: test/token/ERC20WithPermit.test.js [hunk 34]
```diff
@@ -0,0 +1,1102 @@
...
+        it("should emit an approval event", async () => {
+          const deadline = tomorrow
+          const signature = await getApproval(
+            allowance,
+            anotherAccount.address,
+            deadline
+          )
+
+          const tx = await token
+            .connect(anotherAccount)
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:775]: Can we test that anyone (neither owner nor spender) can call `permit` function with a valid signature?

File: test/token/ERC20WithPermit.test.js [hunk 35]
```diff
@@ -0,0 +1,1102 @@
...
+          .burnFrom(initialHolder.address, initialSupply)
+
+        expect(
+          await token.allowance(initialHolder.address, anotherAccount.address)
+        ).to.equal(allowance)
+      })
+    })
+  })
+
+  describe("permit", () => {
```
Comment (nkuba) [inline test/token/ERC20WithPermit.test.js:648]: Can we validate that `nonce` got incremented after each `permit` call?

**Review by nkuba — COMMENTED** [review]
> Let's open a separate issue for test improvements. All tests but permit-related were ported from OpenZeppelin and already reviewed in coverage-pools PR. If we want to do some modifications, let's not hold PRs that wait for this one.

OK. Let's just address comments around the `permit` function which is a new addition.

https://github.com/thesis/solidity-contracts/pull/3#discussion_r678267949
https://github.com/thesis/solidity-contracts/pull/3#discussion_r678272055
https://github.com/thesis/solidity-contracts/pull/3#discussion_r678273496

**Review by nkuba — APPROVED** [review]
